cranelift/codegen/src/ir/pcc.rs

```diff
@@ -224,17 +232,17 @@ pub enum Fact {
 }
 
 /// A bound expression.
-#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
 pub struct Expr {
     /// The dynamic (base) part.
     pub base: BaseExpr,
     /// The static (offset) part.
-    pub offset: i64,
+    pub offset: i128,
```

Comment:
Yeah, I agree; efficiency concerns aside (back to that below) it became necessary because given symbolic expressions, the offset has a domain of... about 3/2 of a `u64`: it sometimes needs to be `i64`, and sometimes needs to be a full `u64`. I've papered over that with 128 bits, though in theory it needs "only" 65. It does feel generally better (less potentially subtly wrong) to use `i128::from(..)` everywhere and be able to perform universal arithmetic, rather than carefully thinking through all the intermedia underflow/overflow cases.

Reply:
Added some notes on offset sizes here as well.